### PR TITLE
[SPARK-13709][SQL] Initialize deserializer with both table and partition properties when reading partitioned tables

### PR DESCRIPTION
**Pull request by liancheng**
## What changes were proposed in this pull request?

When reading partitions of a partitioned Hive SerDe table, we only initializes the deserializer using partition properties. However, for SerDes like `AvroSerDe`, essential properties (e.g. Avro schema information) may be defined in table properties. We should merge both table properties and partition properties before initializing the deserializer.

Note that an individual partition may have different properties than the one defined in the table properties (e.g. partitions within a table can have different SerDes). Thus, for any property key defined in both partition and table properties, the value set in partition properties wins.
## How was this patch tested?

New test case added in `QueryPartitionSuite`.

### CODE REVIEW

**Comment by liancheng** [comment]
This is ready for review.

cc @yhuai @cloud-fan @clockfly

File: sql/hive/src/main/scala/org/apache/spark/sql/hive/TableReader.scala [hunk 1]
```diff
@@ -230,10 +234,21 @@ class HadoopTableReader(
       // Fill all partition keys to the given MutableRow object
       fillPartitionKeys(partValues, mutableRow)
 
+      val tableProperties = relation.tableDesc.getProperties
```
Comment (liancheng) [inline sql/hive/src/main/scala/org/apache/spark/sql/hive/TableReader.scala]: Local variable for serialization.

File: sql/hive/src/test/scala/org/apache/spark/sql/hive/QueryPartitionSuite.scala [hunk 2]
```diff
@@ -65,4 +68,77 @@ class QueryPartitionSuite extends QueryTest with SQLTestUtils with TestHiveSingl
...
+        )
+
+        // Generates Avro data.
+        sql(s"INSERT OVERWRITE TABLE $tempTableName SELECT 1, STRUCT(2, 2.5)")
+
+        // Adds generated Avro data as a new partition to the testing table.
+        sql(s"ALTER TABLE $tableName ADD PARTITION (ds = 'foo') LOCATION '$path/$tempTableName'")
+
+        checkAnswer(
+          sql(s"SELECT * FROM $tableName"),
```
Comment (cloud-fan) [inline sql/hive/src/test/scala/org/apache/spark/sql/hive/QueryPartitionSuite.scala]: can you explain it a bit more how this query fails without your patch?

Comment (liancheng) [inline sql/hive/src/test/scala/org/apache/spark/sql/hive/QueryPartitionSuite.scala]: Yea, when reading data from a partition, the Avro deserializer needs to know the Avro schema defined in the table properties (`avro.schema.literal`). However, originally we only initialize the deserializer using the partition properties, which doesn't contain `avro.schema.literal`. This PR fixes it by merging to sets of properties.

Comment (yhuai) [inline sql/hive/src/test/scala/org/apache/spark/sql/hive/QueryPartitionSuite.scala]: yea, it is a good idea to add comments to explain why this one failed.

Comment (yhuai) [inline sql/hive/src/test/scala/org/apache/spark/sql/hive/QueryPartitionSuite.scala]: drop the table at the end of this test?

Comment (yhuai) [inline sql/hive/src/test/scala/org/apache/spark/sql/hive/QueryPartitionSuite.scala]: oh, nvm. We have withTable.

Comment (liancheng) [inline sql/hive/src/test/scala/org/apache/spark/sql/hive/QueryPartitionSuite.scala]: Yea, sure.

Comment (cloud-fan) [inline sql/hive/src/test/scala/org/apache/spark/sql/hive/QueryPartitionSuite.scala]: it's inside `withTable`, tables will be dropped automatically.

**Comment by cloud-fan** [comment]
LGTM

**Comment by yhuai** [comment]
lgtm

**Comment by yhuai** [comment]
Thanks. Merging to master and branch 2.0.

**Pull request closed without merging**